src/algorithms/PANOC.jl

```diff
@@ -0,0 +1,250 @@
+			    blockaxpy!(sol.FPR_x, sol.x, -1.0, sol.xbar)
+		    else
+			    break
+		    end
+            
+		    Aqxbar = sol.Aq*sol.xbar
+		    gradfq_Aqxbar, fq_Aqxbar = gradient(sol.fq, Aqxbar)
+		    Asxbar = sol.As*sol.xbar
+		    gradfs_Asxbar, fs_Asxbar = gradient(sol.fs, Asxbar)
+		    f_Axbar = fs_Asxbar + fq_Aqxbar
+	    end
+    end
+
+end
+
+################################################################################
+# Iteration
+
+function iterate!(sol::PANOCIterator{I, R, T}, it::I) where {I, R, T}
+
+	xbar = sol.xbar
+	x = sol.x
+
+	if it > 1
+		update!(sol.H, sol.x, sol.x_prev, sol.FPR_x, sol.FPR_x_prev)
+	end
+	A_mul_B!(sol.d, sol.H, 0.0 .- sol.FPR_x) # TODO: not nice
+
+	sol.FPR_x_prev, sol.FPR_x = sol.FPR_x, sol.FPR_x_prev
+	blockset!(sol.x_prev, sol.x)
+
+	FBE_x = sol.FBE_x
+	normFPR_x = sol.normFPR_x
+	C = sol.sigma*sol.gamma*(1.0-sol.alpha)
+	maxit_tau = 10
+	tau = 1.0
+    
+	for it_tau = 1:maxit_tau # TODO: replace/complement with lower bound on tau
+
+		xnew = (1-tau).*xbar .+ tau .* (x.+sol.d)
+	   
+		# calculate new FBE in xnew
+
+		A_mul_B!(sol.Aqx, sol.Aq, xnew)
+		sol.fq_Aqx = gradient!(sol.gradfq_Aqx, sol.fq, sol.Aqx)
+		A_mul_B!(sol.Asx, sol.As, xnew)
+		sol.fs_Asx = gradient!(sol.gradfs_Asx, sol.fs, sol.Asx)
+		blockaxpy!(sol.At_gradf_Ax, sol.As'*sol.gradfs_Asx, 1.0, sol.Aq'*sol.gradfq_Aqx)
+		sol.f_Ax = sol.fs_Asx + sol.fq_Aqx
+
+		# gradient step
+		blockaxpy!(sol.y, xnew, -sol.gamma, sol.At_gradf_Ax)
+		# prox step
+		xnewbar, g_xnewbar = prox(sol.g, sol.y, sol.gamma)
+	    
+		FPR_xnew = xnew .- xnewbar
+		norm_FPRxnew = blockvecnorm(FPR_xnew)
+	    
+		FBE_xnew = sol.f_Ax - blockvecdot(sol.At_gradf_Ax, FPR_xnew) + 
+		           0.5/sol.gamma*norm_FPRxnew^2 + g_xnewbar
+
+		if FBE_xnew <= FBE_x - (C/2)*normFPR_x^2 || it_tau == maxit_tau
+			sol.normFPR_x = norm_FPRxnew
+			sol.FPR_x = FPR_xnew
+			sol.FBE_x = FBE_xnew
+			sol.x, sol.xbar = xnew, xnewbar
+			sol.tau = tau
+			break
+		end
+		tau *= 0.5
+    
+	end
+
+	if sol.adaptive
+		Aqxbar = sol.Aq*sol.xbar
+		gradfq_Aqxbar, fq_Aqxbar = gradient(sol.fq, Aqxbar)
+		Asxbar = sol.As*sol.xbar
+		gradfs_Asxbar, fs_Asxbar = gradient(sol.fs, Asxbar)
+		f_Axbar = fs_Asxbar + fq_Aqxbar
+		for it_gam = 1:100 # TODO: replace/complement with lower bound on gamma
+		    normFPR_x = blockvecnorm(sol.FPR_x)
+		    uppbnd = sol.f_Ax - blockvecdot(sol.At_gradf_Ax, sol.FPR_x) + 0.5/sol.gamma*normFPR_x^2
+		    if f_Axbar > uppbnd + 1e-6*abs(sol.f_Ax)
+			    sol.gamma = 0.5*sol.gamma
+			    blockaxpy!(sol.y, sol.x, -sol.gamma, sol.At_gradf_Ax)
+			    sol.xbar, sol.g_xbar = prox(sol.g, sol.y, sol.gamma)
+			    blockaxpy!(sol.FPR_x, sol.x, -1.0, sol.xbar)
+		    else
+			    break
+		    end
+		
+		    Aqxbar = sol.Aq*sol.xbar
+		    gradfq_Aqxbar, fq_Aqxbar = gradient(sol.fq, Aqxbar)
+		    Asxbar = sol.As*sol.xbar
+		    gradfs_Asxbar, fs_Asxbar = gradient(sol.fs, Asxbar)
+		    f_Axbar = fs_Asxbar + fq_Aqxbar
+		end
+	end
```

Comment:
Can this whole block go at the top of `iterate`? Maybe this way we can avoid having it both in `iterate` and `initialize`.